doc/memos/rdm-draft-weiss-exposing-api-firmware-guidelines.md

```diff
@@ -0,0 +1,141 @@
+- rdm-draft-weiss-exposing-api-firmware-guidelines.md
+- Title: Guidelines for Write Firmware to Expose RIOT APIs
+- Author: Kevin Weiss
+- Status: In-progress
+- Type: Process
+- Date: December 2018
+
+## Abstract
+
+In order to improve the testing and interface tools for RIOT a standard way of
+exposing API calls to a higher level should be defined. This document describes
+a set of guidelines indicating semantics and basic design on the firmware level
+that allow interfaces and tools to be build upon.
+
+## Status
+
+This document is currently under open discussion. Its content is licensed with
+a Creative Commons CC-BY-SA license.
+
+## Terminology
+This memo uses the RFC2119 terminology and the following acronyms:
+API: Application Programming Interface
+cmd: Command
+msg: Message
+
+# 1. Introduction
+The guidelines are used indicate the semantics and design when writing
+firmware to expose APIs through the shell.
+
+_NOTE: It may not be applicable for every API call_
+
+ # 2. Test Firmware Design
+
+The following guidelines are used when writing the C program to be flashed to
+the board.
+
+ ## 2.1. Test Firmware Declarations
+
+- Tests should interact through a shell
+- Commands should be synchronous
+- Tests should have a `get_metadata` command
+- The `get_metadata` command should contain at least board name, and
+application name
+- Tests should have a `echo <random 8 hex digits>` command
+- Test commands should provide similar API/function calls
+- Tests that require `test_<name of test command>` command only if interaction
+with the shell causes inaccuracies
+- Commands that use `test_<name of test command>` should give as much data
+and information as possible to reproduce command and allow higher level
+functionality
+- Asynchronous actions should be hidden with the `test_<name of test command>`
+format
+- There should be a standard interface for every command sent
+[cmd, data, msg, result]
+- [cmd] should only be provided once for each command
+- [cmd] should allow the user to reproduce the step, this may be the
+API/function call with the variables or a descriptive name for a set of calls
+- [msg] should be optional
```

Comment:
Instead of writing so much text, you could roughly specify the grammar of the response:

```
response := message* data* result

message := MESSAGE_CONTENTS EOL

data := <json> EOL

result := "{" "'result'" ":" RESULT_VALUE "}" EOL

EOL := '\n'
MESSAGE_CONTENTS := ([^{].+)?
RESULT_VALUE := (SUCCESS) | (ERROR)
```

Reply:
Can I have both? I will add that in.